src/tnc.rs

```diff
@@ -163,7 +168,33 @@ impl Tnc {
             ConnectConfig::TcpKiss(config) => Box::new(TcpKissTnc::open(&config)?),
             ConnectConfig::LinuxIf(config) => Box::new(LinuxIfTnc::open(&config)?),
         };
-        Ok(Self { imp })
+        Ok(Tnc::new(imp))
+    }
+
+    fn new(imp: Box<dyn TncImpl>) -> Self {
+        let tnc = Tnc {
+            imp,
+            senders: Arc::new(Mutex::new(Vec::new())),
+            buffer: Arc::new(Mutex::new(VecDeque::new())),
+        };
+        {
+            let tnc = tnc.clone();
+            // TODO how to allow this thread to exit?
+            thread::spawn(move || {
+                loop {
```

Comment:
Any idea how we could handle this `TODO` ?

Reply:
Excellent question... apparently I never actually came up with a way to shut down a `Tnc` while we're "stuck" in a call to `receive_frame`. All right I think we can push this into the impls - while we are doing `imp.receive_frame` on our inner thread, the user should be able to call a new `fn shutdown(&self)` on the `Tnc` which calls `imp.shutdown()`. This will destroy the underlying connection (whatever it is) and cause the ongoing call to `receive_frame` to error out. This is similar to `TcpStream` where you can call [`shutdown(Shutdown::Both)` on the transmitting side](https://doc.rust-lang.org/std/net/struct.TcpStream.html#method.shutdown) and tear down the whole thing. This is how the KISS TNC impl currently works.

* For the `TncImpl` trait, add a new `fn shutdown(&self)` and add this to both impls.
    * Ensure the code from the existing `impl Drop` gets run, to close the socket/connection
    * `impl Drop` should instead call `shutdown()`
    * Track when we have shutdown so we don't call it a second time on `Drop`
    * Since we need to call `imp.receive_frame()` and `imp.shutdown()` in parallel they will both require only `(&self)`. I think this is okay, but `linux::sys::socket_close` probably needs the `&mut` specifier removed.

Now our thread will receive an error after the user calls `shutdown()`, and it can terminate gracefully as described in my next comment.

Reply:
I think it has to be `fn shutdown(&mut self)`, just so that we can update a variable to prevent calling shutdown multiple times.

Reply:
IMO (yet another) Mutex for the `is_shutdown` flag is a fair price to pay to make this actually possible.

Reply:
Fixed.